### PR TITLE
Move "listObjectsDescendant" into the S3.Service

### PR DESCRIPTION
**Pull request by melgenek**
There are multiple operators that return an SdkPublisher in the response. One of them is `listObjectsV2Paginator` which returns all the objects instead of one page.

`listObjectsDescendant` is reimplemented based on the streaming operator. The `listAllObjects` is also added into the Service so that the code that is based on ZLayers can inject the Service and use the `listObjectsDescendant`.

`S3ObjectSummary` is extended with additional fields. I have a use case to filter the objects based on the creation time.

### CODE REVIEW

File: src/main/scala/zio/s3/Live.scala [hunk 1]
```diff
@@ -210,8 +214,12 @@ final class Live(unsafeClient: S3AsyncClient) extends S3.Service {
                   )
     } yield ()
 
-  def execute[T](f: S3AsyncClient => CompletableFuture[T]): ZIO[Any, S3Exception, T] =
+  override def execute[T](f: S3AsyncClient => CompletableFuture[T]): ZIO[Any, S3Exception, T] =
     ZIO.fromCompletionStage(f(unsafeClient)).refineToOrDie[S3Exception]
+
+  override def executePublisher[T](f: S3AsyncClient => SdkPublisher[T]): ZStream[Any, S3Exception, T] =
```
Comment (regis-leray) [inline src/main/scala/zio/s3/Live.scala:220]: can you explain why to expose this ?
What is the use case ?

File: src/main/scala/zio/s3/Live.scala [hunk 2]
```diff
@@ -95,6 +95,10 @@ final class Live(unsafeClient: S3AsyncClient) extends S3.Service {
         ).map(S3ObjectListing.fromResponse)
       }
 
+  override def listAllObjects(bucketName: String, prefix: String): ZStream[Any, S3Exception, S3ObjectSummary] =
```
Comment (regis-leray) [inline src/main/scala/zio/s3/Live.scala:98]: by the way the method is already defined

```scala
/**
   * List all descendant objects of a bucket
   * Fetch all objects recursively of all nested directory by traversing all of them
   *
   * @param bucketName name of the bucket
   * @param prefix filter all object identifier which start with this `prefix`
   */
  def listObjectsDescendant(bucketName: String, prefix: String): S3Stream[S3ObjectSummary]
```

**Comment by regis-leray** [comment]
Can you please explain in details,  i dont know what is the problem you are facing.
```
listObjectsDescendant is reimplemented based on the streaming operator. The listAllObjects is also added into the Service so that the code that is based on ZLayers can inject the Service and use the listObjectsDescendant.
```

@melgenek i think it will be a great addition to add the `lastModified` inside `S3ObjectSummary`
```
S3ObjectSummary is extended with additional fields. I have a use case to filter the objects based on the creation time.
```

File: src/main/scala/zio/s3/Live.scala [hunk 3]
```diff
@@ -95,6 +95,10 @@ final class Live(unsafeClient: S3AsyncClient) extends S3.Service {
...
+    executePublisher(_.listObjectsV2Paginator(ListObjectsV2Request.builder().bucket(bucketName).prefix(prefix).build()))
```
Comment (regis-leray) [inline src/main/scala/zio/s3/Live.scala:99]: https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/s3/S3AsyncClient.html#listObjectsV2Paginator-software.amazon.awssdk.services.s3.model.ListObjectsV2Request-

`Returns some or all (up to 1,000) of the objects in a bucket. `

File: src/main/scala/zio/s3/Live.scala (hunk 1 above)
Comment (regis-leray) [reply src/main/scala/zio/s3/Live.scala:220]: i would not expose this method

Comment (melgenek) [reply src/main/scala/zio/s3/Live.scala:220]: My idea of exposing the `executePublisher` is the same as `execute`: the methods that are present in `S3AsyncClient`, but absent in zio-s3, can be used via wrapping the value into ZStream/ZIO by calling `execute*`.
Now there are 3 methods returning the publisher present: `listMultipartUploadsPaginator`, `listObjectsV2Paginator` and `listObjectVersionsPaginator`. I can technically expose them and hide the `executePublisher`. However, in this case, any additional methods on `S3AsyncClient` wouldn't be exposed in any way to the zio-s3 user in case they are added to the `S3AsyncClient`. 

`listAllObjects` was added just to replace the custom implementation `listObjectsDescendant` and as an example usage of the `executePublisher`. The current `listObjectsDescendant` cannot be used in an application that uses ZLayer, because the `listObjectsDescendant` is not present on the layer level. So
```
ZLayer.fromService[S3.Service, MyService]{ s3 => ... }
```
in this situation `s3.listObjectsDescendant` is not present. And I need it. So I added the `listAllObjects` which has the same implementation as `listObjectsDescendant` but is provided by the aws sdk rather than being rewritten in zio-s3.

To sum up this message: I would add the streaming methods that I need (`listMultipartUploadsPaginator` and `listObjectsV2Paginator`) explicitly into the `S3.Service`. And I would also keep the `executePublisher` in order to cover any other methods from aws sdk that can potentially appear in newer versions of sdk.

File: src/main/scala/zio/s3/Live.scala (hunk 3 above)
Comment (melgenek) [reply src/main/scala/zio/s3/Live.scala:99]: The documentation has a not in the bottom 
```
Please notice that the configuration of MaxKeys won't limit the number of results you get with the paginator. It only limits the number of results in each page.
```
I added a test in order to show the difference.
So essentially the `*Paginator` methods in the aws sdk are implemented the same way the `listObjectsDescendant` is implemented in zio-s3: they traverse over all the pages and return them a publisher.

Note 1: while implementing the test I found an issue with the Test s3 implementation. The files that are named like `1001obj/0YM2XhBT5o` so containing `/` in the name cannot be uploaded because the intermediate folders are not created on the disk. I added the operation to create intermediate folders. The querying by prefix also works.

Note 2: I changed the tests to run sequentially. `putObject` creates files and the parallel tests that list files can see this uploaded objects. This behaviour wasn't visible on a couple of uploads, but it is occurring with a lot of files (1000 in the new test).
An alternative solution would be to have a subfolder per test so that the upload/listing operations in parallel tests don't interfere with each other.

File: src/main/scala/zio/s3/Live.scala (hunk 1 above)
Comment (regis-leray) [reply src/main/scala/zio/s3/Live.scala:220]: Oki then the solution is to move all the methods `streamLines(), paginate(), listObjectsDescendant()` inside 
`trait S3.Service`

Comment (regis-leray) [reply src/main/scala/zio/s3/Live.scala:220]: Concerning exposing the publisher, most of the methods in AsyncClient return A `CompletableFuture` and the other methods like `listObjectsV2Paginator` are already implemented or have an equivalent with future.

File: src/test/scala/zio/s3/S3Test.scala [hunk 4]
```diff
@@ -53,25 +54,46 @@ object S3Suite {
...
-        } yield assert(succeed.bucketName)(equalTo(bucketName)) && assert(succeed.objectSummaries)(
+        } yield assert(succeed.bucketName)(equalTo(bucketName)) && assert(
+          succeed.objectSummaries.map(noModificationTimeSummary)
+        )(
           hasSameElements(
             List(
-              S3ObjectSummary(bucketName, "console.log"),
-              S3ObjectSummary(bucketName, "dir1/hello.txt"),
-              S3ObjectSummary(bucketName, "dir1/user.csv")
+              S3ObjectSummary(bucketName, "console.log", FakeModificationTime, 33),
```
Comment (regis-leray) [inline src/test/scala/zio/s3/S3Test.scala:62]: what do you think of

```scala
assert(succeed.bucketName)(equalTo(bucketName)) && assert(
          succeed.objectSummaries.map(s => s.bucketName -> s.key)
        )(
          hasSameElements(
            List(
              (bucketName, "console.log"),
              (bucketName, "dir1/hello.txt"),
              (bucketName, "dir1/user.csv")
            )
          )
        )
```

File: src/test/scala/zio/s3/S3Test.scala [hunk 5]
```diff
@@ -53,25 +54,46 @@ object S3Suite {
...
+              S3ObjectSummary(bucketName, "dir1/user.csv", FakeModificationTime, 181)
             )
           )
         )
       },
       testM("list objects with prefix") {
         for {
           succeed <- listObjects(bucketName, "console", 10)
-        } yield assert(succeed)(
+        } yield assert(succeed.copy(objectSummaries = succeed.objectSummaries.map(noModificationTimeSummary)))(
```
Comment (regis-leray) [inline src/test/scala/zio/s3/S3Test.scala:72]: i would prefer this approach

```scala

assert(succeed)(
          hasField("bucketName", (l: S3ObjectListing) => l.bucketName, equalTo(bucketName)) &&
          hasField("objectSummaries",
              (l: S3ObjectListing) => l.objectSummaries.map(o => o.bucketName -> o.key),
              equalTo(Chunk.single((bucketName, "console.log")))
          )
        )
```

**Review by regis-leray — CHANGES_REQUESTED** [review]
Thanks for your contribution. If you can apply my feedback, that would be great.

File: src/main/scala/zio/s3/Test.scala [hunk 6]
```diff
@@ -122,15 +136,22 @@ object Test {
...
+          _       <-
             refDb.update(db =>
               db + (bucketName + key -> (options.contentType.getOrElse("application/octet-stream") -> options.metadata))
             )
-          _ <- FileChannel
-                 .open(path / bucketName / key, StandardOpenOption.WRITE, StandardOpenOption.CREATE_NEW)
-                 .provide(blocking)
-                 .use(channel => content.foreachChunk(channel.writeChunk))
-        } yield ()).mapError(S3ExceptionUtils.fromThrowable)
+          filePath = path / bucketName / key
```
Comment (regis-leray) [inline src/main/scala/zio/s3/Test.scala:143]: thanks for finding the bug & fixing it

**Comment by melgenek** [comment]
@regis-leray I removed the public method `executePublisher`.
So this pr is about:
- moving the `listObjectsDescendant` into the Service
- reimplementing `listObjectsDescendant` with the aws-sdk built-in operator that does pagination
- fixing the test implementation to be able to create objects that have the key that is separated with "/"

The `listObjectsDescendant` was renamed to `listAllObjects`. It is a bit confusing that the name contains the "descendant" part, because the s3 object listing is returned in alphabetically ascending order.

I am not going to add any additional streaming operators in this pr.

File: src/test/scala/zio/s3/S3Test.scala (hunk 5 above)
Comment (melgenek) [reply src/test/scala/zio/s3/S3Test.scala:72]: I thought about this approach. Why I decided to avoid it is because the `S3ObjectListing` can be extended even further, for example, with the `eTag` field. When using the `hasField` the compilation of the test won't fail. With the current approach the person who adds the additional field is enforced to adjust the test, because it requires all the field to instantiate the case class in the test case.

**Comment by regis-leray** [comment]
@melgenek 

listObjectsDescendant = Returns some or all of the objects in a bucket (by using recursion)
paginate => fetch by pagination list of 
listObjects => Returns some or all (up to 1,000) of the objects in a bucket

i wanted to expose a paginate function for any specific and build on top of it.

i do agree we could replace `listObjectsDescendant` &  `paginate` by using directly `listObjectsV2Paginator` but we will loose the ability to compose operation which i would always prefer, since he gives more power in the end user.

**Comment by regis-leray** [comment]
we could rename `listObjectsDescendant` to listAllObjects`

**Comment by melgenek** [comment]
Hi @regis-leray. Do you have any more comments regarding this pr or can you merge it?

**Comment by regis-leray** [comment]
@melgenek sorry for the long delay and happy new year.

I would prefer not to introduce a new implementation for `listAllObjects` and use existing functions.

```scala
def listAllObjects(bucketName: String, prefix: String): Stream[S3Exception, S3ObjectSummary] =
        ZStream
          .fromEffect(self.listObjects(bucketName, prefix, 1000))
          .flatMap(
            paginate(_).mapConcat(_.objectSummaries)
          )
```

**Comment by melgenek** [comment]
The snippet that you sent doesn't compile, because there is `S3` in the environment. It's possible to move the `paginate` into the `S3.Service` and avoid having the environment this way. Though it's a bit hard to understand the purpose of this method because its meaning is "all pages including the first one when the first one is already provided and it's not empty".

**Comment by regis-leray** [comment]
sorry for the long long delay.
im closing this PR in favor of https://github.com/zio/zio-s3/pull/171.
I took your fixes. Thanks again for your contribution.
